### ISSUE TITLE
Replicating Table 1 Paper Results for 90% Prune Rate

### ISSUE THREAD

**Issue by griffbr** [zhangxin-xd/Dataset-Pruning-TDDS#3]
Hi,

Thanks again for your help on previous issues.

Would you please confirm that the following commands are correct for replicating the `CIFAR-100`, `90%` accuracy result of `54.51`:

```
python train.py --data_path ./data --dataset cifar100 --arch resnet18 --epochs 200 --learning_rate 0.1 --batch-size 100 --dynamics --save_path ./checkpoint/all-dataset
```

```
python importance_evaluation.py --dynamics_path ./checkpoint/all-dataset/npy/ --mask_path ./checkpoint/generated_mask/ --trajectory_len 10 --window_size 5 --decay 0.9
```

```
python train_subset.py --data_path ./data --dataset cifar100 --arch resnet18 --epochs 200 --learning_rate 0.1 --batch-size 128 --save_path ./checkpoint/pruned-dataset --subset_rate 0.9 --mask_path ./checkpoint/generated_mask/data_mask_win5_ep10.npy --score_path ./checkpoint/generated_mask/score_win5_ep10.npy
```

Reason being, I have attempted _all_ of the Table 1 results and find the `30%`-`70%` prune rates are within the +/- reported but the `90%` prune rate has a substantial deviation following the same process.

For more details, here are the logs I generated for:

`CIFAR-100` full dataset (`78.37%` accuracy vs `78.21%` in paper): 
[log_seed_42.txt](https://github.com/user-attachments/files/16178498/log_seed_42.txt)

`CIFAR-100` `30%` prune (`77.46%` accuracy vs `77.56%` in paper):
[log_seed_42.txt](https://github.com/user-attachments/files/16178588/log_seed_42.txt)

`CIFAR-100` `90%` prune (`42.11%` accuracy vs `54.51%` in paper):
[log_seed_42.txt](https://github.com/user-attachments/files/16178642/log_seed_42.txt)

If the above commands are correct, would you please share your logs for the `CIFAR-100` `90%` prune (`54.51%` accuracy) in this thread so I can troubleshoot where this discrepancy is coming from?

Also, how are the `+/-` deviations in Table 1 calculated? I repeated several experiments using this repo and found the training / evaluation process to be deterministic.

![Screenshot 2024-07-11 at 10 06 17 AM](https://github.com/zhangxin-xd/Dataset-Pruning-TDDS/assets/26256043/62d8cac1-0b88-441d-baac-ffc892a95a05)

Any guidance is much appreciated!

**Comment by griffbr** [comment]
As a quick followup, I noticed the repo note that "*For an aggressive pruning rate, setting a smaller batch size will lead to better performance." and then from your paper:
![Screenshot 2024-07-11 at 10 23 52 AM](https://github.com/zhangxin-xd/Dataset-Pruning-TDDS/assets/26256043/0775b4d2-448e-49b0-9173-1ab81570ae4c)

Also from supplemental material:
![Screenshot 2024-07-11 at 10 29 58 AM](https://github.com/zhangxin-xd/Dataset-Pruning-TDDS/assets/26256043/d364c92a-3989-4917-b425-cbff4b1fb71c)


Apologies for missing this detail before. I will rerun `80%`, `90%` with the reduced batch size and provide an update once finished. However, my result of `CIFAR-100` `90%` prune `128` batch is _still_ substantially lower than the paper's corresponding result in Table 6 (`42.11%` accuracy vs `51.32%` in paper). Would be great to know if you have any thoughts or suggestions on this.

In the meantime, it would be great to understand how the `+/-` deviations are calculated. Thanks!

**Comment by griffbr** [comment]
Using the reduced batch size, I replicated all of the paper's Table 1 results _except for_ the 90% prune on CIFAR-100.

Here is my train / eval log for `CIFAR-100` `90%` prune `32` batch (`51.83%` accuracy vs `54.51%` in paper):
[log_seed_42.txt](https://github.com/user-attachments/files/16181396/log_seed_42.txt)

_All_ other replicated CIFAR results are either within or _fairly_ close to being within the `+/-` paper results.

## Open questions:

**1.** How did you specifically calculate the `+/- 0.22` for the `CIFAR-100` `90%` prune result in Table 1? 

For example, did this only involve changing the seed of `train_subset.py`, or did this include recalculating the original loss gradients of `train.py` from a different seed as well? How many times did you repeat this experiment?

**2.** Would you share the `CIFAR-100` `90%` prune train / eval logs from the paper here so I can understand this better?

Thanks again for the paper, repo, and all of your help!

**Comment by zhangxin-xd** [comment]
Question 1:

We conducted five experiments for each pruning rate and then calculated the standard deviation and mean.

Question 2:

Please allow me to verify it on my end.  I will get back to you as soon as I have the results.

**Comment by zhangxin-xd** [comment]
Hi Brent, I suggest using a larger weight decay parameter for this aggressive pruning rate.

**Comment by griffbr** [comment]
Thanks for your help!

1. "We conducted five experiments for each pruning rate and then calculated the standard deviation and mean." 
Did you repeat training on the full dataset for gradient-based selection each of the 5 times (`train.py`) or only the pruned subset training (`train_subset.py`)?

2. "Please allow me to verify it on my end. I will get back to you as soon as I have the results." Thank you so much!

3. "I suggest using a larger weight decay parameter for this aggressive pruning rate."

As best as I can tell, I am already using the weight decay recommended by the paper and repo:
![Screenshot 2024-07-12 at 10 01 45 AM](https://github.com/user-attachments/assets/3770abba-10b8-440a-9b9a-c34d83ac2246)

```
parser.add_argument('--momentum', type=float, default=0.9, help='Momentum.')
parser.add_argument('--decay', type=float, default=0.0005, help='Weight decay (L2 penalty).')
```

What settings were used to generate the paper's `CIFAR-100` `90%` prune results in Table 1?

**Comment by zhangxin-xd** [comment]
Hello, please find the attached training log for your reference.
[log_seed_42.txt](https://github.com/user-attachments/files/16233567/log_seed_42.txt)

**Comment by griffbr** [comment]
Thanks for sharing the above log. Using the weight decay setting from that log (`'decay': 0.0007`), I get an accuracy of `54.03%` and `53.19%` on two full separate runs (that is, repeating the full process of `train.py`, `importance_evaluation.py`, `train_subset.py`). Here are the logs of those two runs for your reference:
[log_seed_42.txt](https://github.com/user-attachments/files/16238913/log_seed_42.txt)
[log_seed_42.txt](https://github.com/user-attachments/files/16238927/log_seed_42.txt)

However, none of my experiments with your most recently proposed settings nor the log you shared (`54.22%`) are within the paper's reported `54.51% +/- 0.22` for `CIFAR-100` `90%` prune rate in Table 1.

What do you think is causing this difference? 

Were any other settings changed to generate the paper's `CIFAR-100` `90%` prune rate in Table 1?

_Would you please_ share the `CIFAR-100` `90%` prune train / eval logs from the paper here so I can understand this better?

Thanks again for all of your help.
